formula1_race_analysis/q1_session_analyzer.py

```diff
@@ -0,0 +1,161 @@
+    """
+    Parses a log file to extract driver timestamps.
+    Raises CustomInvalidFormatDataError when data in file is not in the correct format.
+    """
+    driver_timestamp = {}
+    file_content = read_file_content(filepath)
+    for line in file_content:
+        log_info = line.strip("\n")
+        driver_id = log_info[:3]
+        timestamp = log_info[3:]
+        if not driver_id.isalpha() or not driver_id or not timestamp:
+            raise CustomInvalidFormatDataError(f"Error! Incorrect data format in file: {filepath}.")
+        driver_timestamp[driver_id] = timestamp
+    return driver_timestamp
+
+
+def calculate_race_duration(database: list[Driver]) -> None:
+    """
+    Calculates the race duration for each driver in the database and updates the database.
+    Raised CustomMissedTimestampError is raised when a race start or end time isn't provided.
+    Raised CustomInvalidRaceTimeError is raised when the driver' race start time exceeds than the race's end time.
+    """
+    for driver in database:
+        if driver.start_time is None or driver.end_time is None:
+            raise CustomMissedTimestampError(
+                f"Error! Missing timestamp for driver: '{driver.driver_name}'. Start or end time is not provided.",
+            )
+        start_time_in_seconds = convert_timestamp_to_seconds(driver.start_time)
+        end_time_in_seconds = convert_timestamp_to_seconds(driver.end_time)
+
+        race_duration_seconds = end_time_in_seconds - start_time_in_seconds
+        if start_time_in_seconds > end_time_in_seconds:
+            raise CustomInvalidRaceTimeError(
+                f"Race time error for driver: '{driver.driver_name}'. Start time is greater than end time.",
+            )
+        driver.race_duration = race_duration_seconds
+
+
+def convert_timestamp_to_seconds(timestamp: str) -> float:
+    """
+    Converts the timestamp format "YYYY-MM-DD_HH:MM:SS.sss" to a total number of seconds.
+    If the timestamp isn't in the correct format, a CustomInvalidFormatDataError raises.
+    """
+    try:
+        dt = datetime.strptime(timestamp, "%Y-%m-%d_%H:%M:%S.%f")
+        hours = dt.hour
+        minutes = dt.minute
+        seconds = dt.second
+        milliseconds = dt.microsecond / 1000000
+        total_seconds = hours * SECONDS_PER_HOUR + minutes * SECONDS_PER_MINUTE + seconds + milliseconds
+    except ValueError as error:
+        raise CustomInvalidFormatDataError(
+            f"The timestamp format: '{timestamp}' is incorrect. Expected format: YYYY-MM-DD_HH:MM:SS.sss",
+        ) from error
+    return total_seconds
+
+
+def get_q1_driver_position(database: list) -> list:
+    """
+    This function sorts the input list of drivers by their race duration in ascending order.
+    Then assigns qualifying positions (Q1) to drivers based on their race durations.
+    """
+    sorted_driver_database = sorted(database, key=lambda driver: driver.race_duration)
```

Comment:
race_diration is for everyone 20 or whatever laps. It's either wrong logic or wrong naming of the attribute